Synonyms APIs - Synonyms Sets documentation

carlosdelest opened the pull request:
Add documentation for top-level synonyms sets APIs.

There is more work to come, but I'd like an initial review for the top level APIs so I can apply it to the next APIs to document 😉 

Screenshots:

![localhost_8000_guide_synonym-management-apis html](https://github.com/elastic/elasticsearch/assets/6339205/386fae2c-764a-4dc7-9eaa-51b6cb917d81)

![localhost_8000_guide_put-synoyms-set html](https://github.com/elastic/elasticsearch/assets/6339205/29f096f5-9dc3-4397-ae48-37d709cfaafc)

![localhost_8000_guide_get-synoyms-set html](https://github.com/elastic/elasticsearch/assets/6339205/4b3fcc62-8f2a-4e74-87a1-78f903c5b8ce)

![localhost_8000_guide_list-synoyms-set html](https://github.com/elastic/elasticsearch/assets/6339205/0914584f-58e7-4abd-a122-d9897ae0d5cf)

![localhost_8000_guide_delete-synoyms-set html](https://github.com/elastic/elasticsearch/assets/6339205/18b9c273-bf95-40e0-9044-7bc9c2a62b3f)

Commit:
Include put and overview docs

Commit:
Add Get Synonyms Set

Commit:
Delete operation

Commit:
List synonyms sets

Commit:
Add privileges and internal users doc

Commit:
Fix API specs URLs

elasticsearchmachine commented:
Pinging @elastic/es-docs (Team:Docs)

kderusso commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+==== {api-prereq-title}
+
+* Requires the `manage_search_synonyms` cluster privilege.
+* You can only delete a Synonym Set that is not in use by any <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>> that are included in an analyzer. See <<delete-synonym-set-analyzer-requirements>> for more information.
```
I _think_ we want to lowercase `synonym(s) sets`. Also, I think we want to be consistent with pluralization throughout our documentation and refer to it as a `synonyms set`? (This comment applies throughout the PR)

```suggestion
* You can only delete a synonyms set that is not in use by any <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>> that are included in an analyzer. See <<delete-synonym-set-analyzer-requirements>> for more information.
```

carlosdelest replied:
I'm happy to change Synonyms Sets to lowercase. I was taking Search Applications example for this one - I'd like to have guidance from the docs team as I'm not sure of what should be the correct form 👍 

I've fixed the plurals, thanks for catching it!

kderusso replied:
The general guidance is that product names should be Upper Cased and "things" should be lower cased. Search Applications is a weird case because it refers to the branded product name (Search Applications) but at the end of the day it's a search application. 

Now that I've confused you more I'll cede to the docs team 😂

chriscressman replied:
My official docs opinion :smile: : plural and lowercase

carlosdelest replied:
Thank you both! 🙇

carlosdelest replied:
I'll create a separate PR for doing code changes for being consistent with the plurals - I'd prefer to keep this PR as focused on docs as possible

kderusso commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+==== {api-response-codes-title}
+
+`400`::
+The `synonyms_set` identifier was not provided, or the Synonyms Set can't be deleted because it does not meet the <<delete-synonym-set-analyzer-requirements>>.
```
```suggestion
The `synonyms_set` identifier was not provided, or the Synonyms Set can't be deleted because it does not meet the <<delete-synonym-set-analyzer-requirements, specified requirements>>.
```

carlosdelest replied:
that looks better!

kderusso commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+Synonyms Sets can be used in  <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>>.
+These synonym filters can be used as part of <<search-analyzer, search analyzers>>.
+
+Analyzers need to be loaded when an index is restored (such as when a node starts, or the index becomes open).
```
Worth linking to https://www.elastic.co/guide/en/elasticsearch/reference/current/search-analyzer.html here?

carlosdelest replied:
it can be a bit redundant as it's linked in the previous line 🤔 . What does the docs team think about repeating links on adjacent paragraphs, should we do that?

kderusso replied:
Hmm that's a fair point 👍

kderusso commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+Analyzers need to be loaded when an index is restored (such as when a node starts, or the index becomes open).
+Even if the analyzer is not used on any field mapping, it still needs to be loaded on the index recovery phase.
+
+In case analyzers can't be loaded, the index would not be made available and the cluster status would become <<red-yellow-cluster-status,red or yellow>> as index shards would not be available.
```
Not 100% sure on this wording, but it seems a bit more concise? 

```suggestion
If any analyzers cannot be loaded, the index becomes available and the cluster status becomes <<red-yellow-cluster-status,red or yellow>> as index shards are not available.
```

carlosdelest replied:
I like that! Did a minor correction and updated it 💯

kderusso commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+The following example deletes a Synonyms Set called `my-synonyms-set`:
+
+[source,console]
+----
```
You probably want to add some tests here, which will include some setup (to create the index that will be deleted). [These instructions](https://github.com/elastic/elasticsearch/tree/main/docs#test-code-snippets) provide a good getting started point.

carlosdelest replied:
Nice! I'll give them a go, thanks!

kderusso commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,150 @@
+`<synonyms_set>`::
+(Required, string)
+Synonyms set identifier to create.
+This identifier will be used by other Synonyms Management APIs to manage the synonyms set.
```
Link to the top level APIs page here?

carlosdelest replied:
Nice!

kderusso reviewed (commented):
Nice work so far! Only a couple of comments.

carlosdelest commented on docs/reference/synonyms/apis/list-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,67 @@
+Retrieves a summary of all defined Synonyms Sets.
+
+This API allows to retrieve the total number of Synonyms Sets defined.
+For each Synonyms Set, its identifier and the total number of rules defined is returned.
```
💯

Commit:
PR feedback

mayya-sharipova commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,31 @@
+
+---
+
+Synonyms Management APIs let you create and manage synonyms sets.
```
May be we should explain what is a synonyms set with some introduction, with something like this:

"In the past there were only two ways to manage synonyms: providing synonyms inline while defining an analyzer or providing synonyms in a file.  Both these methods have their drawbacks, the former could lead to the out-blown mappings, while the later requires keep synonym files consistent across all nodes.

Synonym Management API is a new way to define and manage synonyms through an API where synonyms will be stored in an internal system index.  A number of synonyms can be grouped in a "synonyms set", and you can have as many sets as needed"....

carlosdelest replied:
I like this, but I think this should be part of a blog post explaining the past situation vs the present one. I'm not sure about including the reasons for an API to be created, and talking about "new" might get stale in a few releases from now...

leemthompo replied:
I think we can still extract the key points from @mayya-sharipova's suggestion, without "time stamping" the content. For example, we can still outline the relative merits of this new API approach, like how it avoids problems from existing approaches. :)

kderusso replied:
I like the idea of providing a value statement here, and then saving the full in-depth description for a blog post. 👍

carlosdelest replied:
Thanks for the feedback! How about this? It's a small rephrasing from Mayya's statement that provides the value proposition and the comparison to other methods. LMKWy'allT!

> Synonym Management API provides a way to define and manage synonyms through an API where synonyms are stored in an internal system index. Related synonyms can be grouped in a "synonyms set". As many synonyms sets can be created as you need for managing synonyms.
> 
> This provides an alternative to the other synonyms definition methods:
> - Inline synonyms in an analyzer definition. This can lead to out-blown mappings.
> - Synonyms file. It requires to keep synonyms files consistent across all nodes.

kderusso replied:
WDYT about these tweaks?

> The synonyms management API provides a convenient way to define and manage synonyms in an internal system index. Related synonyms can be grouped in a "synonyms set". As many synonyms sets can be created as you need for managing synonyms.
> 
> This provides an alternative to the other synonyms definition methods:
> 
> Defining inline synonyms in an analyzer definition, which can lead to [mappings explosion](https://www.elastic.co/guide/en/elasticsearch/reference/current/mapping-explosion.html) and impact performance 
> Manually uploading a synonyms file and managing consistency across all nodes


On a side note we should probably also update [the synonym token filter doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) to reference the API as well

carlosdelest replied:
Nice! I've put all this in the blender and updated the PR. Thanks for your suggestions!

> On a side note we should probably also update [the synonym token filter doc](https://www.elastic.co/guide/en/elasticsearch/reference/current/analysis-synonym-tokenfilter.html) to reference the API as well

For sure - that's something I want to tackle separately, as I'd like to check how to reorg the different synonyms definition methods, potentially as a separate guide. I'll ask for the doc teams help on this 👍

mayya-sharipova commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,31 @@
+
+Synonyms sets can be used to configure <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>>.
+
+Once a synonyms set is used in a synonym token filter, these token filters can be used as part of a <<search-analyzer,search analyzers>>.
```
can be used => must be used or "can only be used"

Synonym token filter can be used  only in search analyzers ...

carlosdelest replied:
I'm rephrasing this as "are used as part", to leave the "can" out 👍

mayya-sharipova commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+==== {api-prereq-title}
+
+* Requires the `manage_search_synonyms` cluster privilege.
+* You can only delete a Synonyms Set that is not in use by any <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>> that are included in an analyzer. See <<delete-synonym-set-analyzer-requirements>> for more information.
```
I think we can say this shortly: 

You can only delete a Synonyms Set that is not in use by any index analyzer.

carlosdelest replied:
Changed!

mayya-sharipova commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+
+[discrete]
+[[delete-synonym-set-analyzer-requirements]]
+==== Delete Synonyms Set analyzer requirements
```
I am not sure if we need all these details and the whole [[delete-synonym-set-analyzer-requirements]] section.

carlosdelest replied:
I think it's important to explain why users can't delete a synonyms set. It took me a while to understand it, and other members of the team as well.

Maybe it's a bit too much technical detail, but I would err on the side of providing more information that justifies the decision.

mayya-sharipova commented on docs/reference/synonyms/apis/delete-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,91 @@
+----
+// TEST[skip:TBD]
+
+In order to remove a Synonyms Set that is being used in a token filter used by an analyzer, the only option would be to remove the index that contains the analyzer.
```
This point is very important, and must be included.

Also we can shorten things: "that is being used in a token filter used by an analyzer," => "used by an analyzer"

carlosdelest replied:
This is more direct, adding that.

mayya-sharipova commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,31 @@
+
+The following APIs can be used to manage synonyms sets:
+
+* <<put-synoyms-set>>
```
<<put-synoyms-set>> => <<put-synonyms-set>> and others

carlosdelest replied:
thanks!

chriscressman commented on docs/reference/synonyms/apis/list-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,67 @@
+
+Retrieves a summary of all defined Synonyms Sets.
+
+This API allows to retrieve the total number of Synonyms Sets defined.
```
```suggestion
This API allows you to retrieve the total number of Synonyms Sets defined.
```

chriscressman commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,150 @@
+
+`<body>`::
+(Required, object)
+Contains the definition of the synonym rules that conform the synonyms set:
```
```suggestion
Contains the definition of the synonym rules for this synonyms set:
```

chriscressman commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,150 @@
+[discrete]
+==== Analyzer reloading
+When an existing synonyms set is updated, the <<search-analyzer, search analyzers>> that use the synonyms set are reloaded automatically for all indices.
+This would be equivalent to invoke <<indices-reload-analyzers>> for all indices that use the synonyms set.
```
```suggestion
This would be equivalent to invoking <<indices-reload-analyzers>> for all indices that use the synonyms set.
```

chriscressman commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,31 @@
+[[synonym-management-apis]]
+== Synonym Management APIs
```
```suggestion
== Synonyms APIs
```
Plural to match API path. Removed "management" for consistency with other CRUD APIs ("management" is implied). This change would cascade through to a few other places that I didn't mark.

carlosdelest replied:
Changing that, thanks Chris!

mayya-sharipova commented on docs/reference/synonyms/apis/get-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,78 @@
+  ]
+}
+----
+// TEST[skip:TBD]
```
What is the reason we are skipping running these doc tests?

carlosdelest replied:
I hadn't started working on tests yet - I'll give it a go

mayya-sharipova commented on docs/reference/synonyms/apis/list-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,67 @@
+beta::[]
+
+++++
+<titleabbrev>List Synonyms Set</titleabbrev>
```
"Synonyms Set" => "Synonyms Sets"

mayya-sharipova commented on docs/reference/synonyms/apis/list-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,67 @@
+[[list-synonyms-set-example]]
+==== {api-examples-title}
+
+The following example retrieves all defined Synonyms Set:
```
Set => Sets

mayya-sharipova commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,150 @@
+
+`<body>`::
+(Required, object)
+Contains the definition of the synonym rules that conform the synonyms set:
```
What does it mean  "synonym rules that conform the synonyms set:"?

carlosdelest replied:
Chris had a better suggestion: "Contains the definition of the synonym rules for this synonyms set"

carlosdelest committed:
Apply suggestions from code review

Co-authored-by: Chris Cressman <chris@chriscressman.com>

Commit:
PR review comments

Commit:
PUT tests

Commit:
DELETE tests

Commit:
List tests

Commit:
Merge branch 'main' into carlosdelest/synonyms-api-docs

# Conflicts:
#	x-pack/docs/en/security/authorization/privileges.asciidoc

carlosdelest commented:
@kderusso , @mayya-sharipova , I added working tests to the docs 👍

Commit:
Fix test

mayya-sharipova commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,215 @@
+// TODO link to synonym rules APIs
+The identifier associated to the synonym rule, that can be used to manage individual synonym rules via Synonym Rules APIs.
+In case a synonym rule id is not specified, an identifier will be created automatically by the API.
+
```
by the API =>  by {es} 

We usually say "by Elasticsearch"

mayya-sharipova reviewed (approved):
@carlosdelest Great work as usual, LGTM! +1 to merge after adding some introduction to the Synonyms Management API doc, that also explains what is "synonyms set".

Commit:
Add value proposition

Commit:
PR comments

carlosdelest commented:
@leemthompo @chriscressman @kderusso , anything else that stands out as wrong here? 

I'm mostly looking for things not to do in the following PRs for the remaining APIs 😉

leemthompo commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,36 @@
+
+Synonyms sets can be used to configure <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>>.
+
+Once a synonyms set is used in a synonym token filter, these token filters are used as part of a <<search-analyzer,search analyzers>>.
```
```suggestion
When synonyms sets are used in synonym token filters, filters are applied as part of the <<search-analyzer,search analyzer>> analysis process.
```

Just trying to make this clearer, not sure if it helps though :)

leemthompo commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,36 @@
+
+Once a synonyms set is used in a synonym token filter, these token filters are used as part of a <<search-analyzer,search analyzers>>.
+You can dynamically change the synonyms used at search time using these APIs to update the synonyms sets.
+Your search results will immediately reflect the synonyms set change.
```
```suggestion
Your search results will immediately reflect the synonyms set changes.
```

Plurals get confusing here :)

leemthompo commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,36 @@
+You can dynamically change the synonyms used at search time using these APIs to update the synonyms sets.
+Your search results will immediately reflect the synonyms set change.
+
+The following APIs can be used to manage synonyms sets:
```
```suggestion
Use the following APIs to manage synonyms sets:
```

leemthompo reviewed (approved):
LGTM writing wise, just a couple of teeny essentially grammatical nits.

carlosdelest committed:
Apply suggestions from code review

Co-authored-by: Liam Thompson <32779855+leemthompo@users.noreply.github.com>

carlosdelest committed:
Update synonyms-apis after rewording

Commit:
Add cluster stats docs

abdonpijpelink commented on docs/reference/rest-api/index.asciidoc:
```diff
@@ -95,6 +95,7 @@ include::{xes-repo-dir}/rest-api/security.asciidoc[]
 include::{es-repo-dir}/snapshot-restore/apis/snapshot-restore-apis.asciidoc[]
 include::{es-repo-dir}/slm/apis/slm-api.asciidoc[]
 include::{es-repo-dir}/sql/apis/sql-apis.asciidoc[]
+include::{es-repo-dir}/synonyms/apis/synonyms-apis.asciidoc[]
```
Don't forget to add it to the list too (after line 51).

carlosdelest replied:
Thanks for the catch! 💯

abdonpijpelink commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,215 @@
+[[put-synonyms-set]]
+=== Put synonyms set
```
```suggestion
=== Create or update synonyms set
```

abdonpijpelink replied:
The convention has been to say "create or update" instead of "put" (see https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-put-template.html for example)

abdonpijpelink commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,36 @@
+Create as many synonym sets as you need.
+
+This provides an alternative to:
+- Defining inline synonyms in an analyzer definition, which impacts mapping size and can lead to performance issues.
```
```suggestion

- Defining inline synonyms in an analyzer definition, which impacts mapping size and can lead to performance issues.
```

abdonpijpelink replied:
Without this newline the list doesn't render correctly.

carlosdelest replied:
Good catch, thanks! 👍

abdonpijpelink commented on docs/reference/synonyms/apis/synonyms-apis.asciidoc:
```diff
@@ -0,0 +1,36 @@
+Synonyms sets can be used to configure <<analysis-synonym-graph-tokenfilter,synonym graph token filters>> and <<analysis-synonym-tokenfilter,synonym token filters>>.
+These filters are applied as part of the <<search-analyzer,search analyzer>> analysis process.
+
+You can dynamically change the synonyms used at search time using these APIs to update the synonyms sets.
```
```suggestion
You can use these APIs to dynamically update synonyms sets used at search time.
```

abdonpijpelink replied:
Just a suggestion. The original sentence is hard to read.

abdonpijpelink commented on docs/reference/synonyms/apis/put-synonyms-set.asciidoc:
```diff
@@ -0,0 +1,215 @@
+Requires the `manage_search_synonyms` cluster privilege.
+
+[[put-synonyms-set-path-params]]
+==== {api-path-parms-title}
```
I think it would be better to have a section for query parameters and a section for the request body here. See https://www.elastic.co/guide/en/elasticsearch/reference/master/indices-put-template.html#put-index-template-api-path-params for example.

carlosdelest replied:
Thanks for the link, changing it!

abdonpijpelink reviewed (commented):
I've left a few suggestions.

Commit:
Revert changes to cluster-stats API, save them for separate PR

carlosdelest committed:
Apply suggestions from code review

Co-authored-by: Abdon Pijpelink <abdon.pijpelink@elastic.co>

Commit:
PR review comments

Commit:
Fix create or update

carlosdelest commented:
@elasticsearchmachine run elasticsearch-ci/part-3

carlosdelest commented:
I'm merging this as it already has some approvals. Happy to continue addressing feedback on https://github.com/elastic/elasticsearch/pull/98064.

Thanks y'all!
